### ISSUE TITLE
[3.4] Memory leaking using RpcUnreliableId/RpcId

### ISSUE THREAD

**Issue by git2013vb** [godotengine/godot#55086]
### Godot version

v3.4.stable.mono.official [206ba70f4]

### System information

Debian 11

### Issue description

This issue is related to #55027 
In ln client side I use `RpcUnreliableId(Network.ServerID, "ReceivePlayerState", transform);` or` RpcId(Network.ServerID, "ReceivePlayerState", transform);` to send a Dictionary to the server
and I have the same problem as mentioned in the other issue.
![image](https://user-images.githubusercontent.com/13235119/142416724-541d83b6-5ae4-43bb-a10f-eaeaeff74a9a.png)

If I comment the rpc instruction in client it work fine (no leaking)
![image](https://user-images.githubusercontent.com/13235119/142417025-b74d3f86-175b-460c-a936-21c6374c52b5.png)

in server before and after comment rpc instruction:
![image](https://user-images.githubusercontent.com/13235119/142417170-d5fa4846-dbbd-4d51-ae2f-85220a874976.png)

### Steps to reproduce

na

### Minimal reproduction project

na

**Comment by Calinou** [comment]
Try using manual object `Dispose()` to free memory manually.

**Comment by git2013vb** [comment]
@Calinou  Thank you for the tip.
I never used `Dispose()` before. I tried to set a timer (10 sec) in main node and call `Dispose()`
I see no changes, After the first signal call I have this error:

```
ERROR: Error calling method from signal 'timeout': 'Node::_on_Dispose_timeout': Method not found.
   at: emit_signal (core/object.cpp:1236)
```

From some c# tutorials I understand `Dispose()` is used in destructors classes. In my case I use physical process to call rpc. My guess was to set a timer do do the job. I'm missing something? It have to set in different way?

I tried to dispose the network connection but seems not so immediate to me to understand how it work :( 

In case of `GetTree().NetworkPeer.Dispose();` (my network in my case) I have this error at first next network call:

```
Unhandled Exception:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'Godot.NetworkedMultiplayerENet'.
  at Godot.Object.GetPtr (Godot.Object instance) [0x0001b] in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Core/Object.base.cs:43 
  at Godot.NetworkedMultiplayerPeer.GetConnectionStatus () [0x00001] in /root/godot/modules/mono/glue/GodotSharp/GodotSharp/Generated/GodotObjects/NetworkedMultiplayerPeer.cs:166 
  at DesertEdge.Singleton.GameServer.DetermineLatency () [0x00001] in /home/vale/godot/DesertEdgeSharp/client/autoload/GameServer.cs:83 
```
I think I misunderstood how `Dispose()` work. Atm Docs not help me either :(

**Comment by kleonc** [comment]
@git2013vb Please upload a reproduction project. If needed server+client and list of steps needed to get what you call "a leak". Otherwise can't confirm/deny whether there's really some leak, or point out the cause of the increasing memory usage.

> ps.: Why do you think its not a bug?

Just my opinion:
I don't know whether there's a bug. Increasing static memory over time doesn't mean there's a leak. And even if there's a leak, it doesn't mean the leak is in the engine itself. It still might be caused by the user.
However, in most cases reports about leaks in C#/Mono are caused by people not realizing how GC works. So according to that (and [your previous report](https://github.com/godotengine/godot/issues/55027)) I'd also assume there's no leak whatsoever. But again, can't confirm that suspiction without a way to reproduct your issue.

**Comment by git2013vb** [comment]
@kleonc I will try to prepare a client server reproduction project asap :) i think too that is the case now..

**Comment by git2013vb** [comment]
Update: I found the problem (finally ...)

I was using  

`JsonConvert.SerializeObject(worldState, Formatting.Indented);`  where `worldstate` is a Dictionary

in rpc  to display the current Dictionary value on screen.

Than give me memory leak.

Thanks everyone :)

Close it

**Comment by kleonc** [comment]
Since `3.4.beta3.mono` there's a leak when enumerating `Dictionary` using foreach (see #56180). Maybe `JsonConvert.SerializeObject` is internally enumerating `Dictionary` like that and it was the cause?
You could check if it leaks in [`3.4.beta2.mono`](https://downloads.tuxfamily.org/godotengine/3.4/beta2/mono/). If it doesn't leak then this issue is probably a duplicate of #56180 which is already fixed (it's fixed in `master` but not yet included in the releases; I'd say the fix should be included in the next bugfix release (`3.4.3`) and in the next `3.x` release (`3.5`)).

**Comment by git2013vb** [comment]
> > Update: I found the problem (finally ...)
> > I was using
> > `JsonConvert.SerializeObject(worldState, Formatting.Indented);` where `worldstate` is a Dictionary
> > in rpc to display the current Dictionary value on screen.
> > Than give me memory leak.
> > Thanks everyone :)
> > Close it
> 
> Since `3.4.beta3.mono` there's a leak when enumerating `Dictionary` using `foreach` (see #56180). Maybe `JsonConvert.SerializeObject` is internally enumerating `Dictionary` like that and it was the cause? You could check if it leaks in [`3.4.beta2.mono`](https://downloads.tuxfamily.org/godotengine/3.4/beta2/mono/). If it doesn't leak then this issue is probably a duplicate of #56180 which is already fixed (it's fixed in `master` but not yet included in the releases; I'd say the fix should be included in the next bugfix release (`3.4.3`) and in the next `3.x` release (`3.5`)).

I think its possible what you mentioned. 
Because I tried to use System.Text.Json instead and I have the same problem.. 

Unless I convert The dictionary into a `string` with `ToString()`  (which I will lose the formatting) . 

I will wait until next update and I will update this tread

**Comment by mhilbrunner** [comment]
The above issue was fixed for Godot 3.5 and 3.4.3. Please ping us for a reopen if you can reproduce it with those versions or a current build.

**Comment by git2013vb** [comment]
Thanks,
I checked and now it work properly.
Turns out that was a bug :)
